compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -868,12 +875,72 @@ public BType setLiteralValueAndGetType(BLangLiteral literalExpr, BType expType,
+            byte[] byteArray = convertToByteArray(literalExpr);
+            literalType = new BArrayType(symTable.byteType, null, byteArray.length,
+                    BArrayState.CLOSED);
         }
 
         return literalType;
     }
 
+    private byte[] convertToByteArray(BLangLiteral literalExpr) {
+        String[] elements = getLiteralTextValue(literalExpr.toString());
+        if (elements[0].contains(BASE_16)) {
+            return hexStringToByteArray(elements[1]);
+        }
+        return Base64.getDecoder().decode(elements[1].getBytes(StandardCharsets.UTF_8));
+    }
+
+    private static byte[] hexStringToByteArray(String base16String) {
+        int arrayLength = base16String.length();
+        byte[] byteArray = new byte[arrayLength / 2];
+        for (int i = 0; i < arrayLength; i += 2) {
+            byteArray[i / 2] = (byte) ((Character.digit(base16String.charAt(i), 16) << 4)
+                    + Character.digit(base16String.charAt(i + 1), 16));
+        }
+        return byteArray;
+    }
+
+    private String[] getLiteralTextValue(String literalExprText) {
+        String nodeText = literalExprText.replace("\t", "").replace("\n", "").replace("\r", "")
+                .replace(" ", "");
+        String[] result = new String[2];
+        result[0] = nodeText.substring(0, nodeText.indexOf('`'));
+        result[1] = nodeText.substring(nodeText.indexOf('`') + 1, nodeText.lastIndexOf('`'));
+        return result;
+    }
+
+    private BType checkByteArrayType(BLangLiteral literalExpr, BType literalType, AnalyzerData data) {
+        byte[] byteArray = convertToByteArray(literalExpr);
+        if (literalExpr.expectedType.tag == TypeTags.ARRAY) {
+            return checkByteArrayCompatibility(
+                    literalExpr, (BArrayType) literalExpr.expectedType, (BArrayType) literalType, byteArray, data);
+        }
+        return types.checkType(literalExpr, literalType, data.expType);
+    }
+
+    private BType checkByteArrayCompatibility(BLangLiteral literalExpr, BArrayType arrayType, BArrayType literalType,
+                                              byte[] byteArray, AnalyzerData data) {
+        if (arrayType.state == BArrayState.INFERRED) {
+            arrayType.size = byteArray.length;
+            arrayType.state = BArrayState.CLOSED;
+        }
+
+        if (arrayType.state != BArrayState.OPEN && byteArray.length != arrayType.size) {
+            dlog.error(literalExpr.pos, DiagnosticErrorCode.MISMATCHING_ARRAY_LITERAL_VALUES, arrayType.size,
+                    byteArray.length);
+            return symTable.semanticError;
+        }
+
+        if (Symbols.isFlagOn(arrayType.flags, Flags.READONLY)) {
+            boolean isValidType =
+                        types.checkType(literalExpr.pos, symTable.byteType, arrayType.eType,
+                                DiagnosticErrorCode.INCOMPATIBLE_TYPES) != symTable.semanticError;
+            return isValidType ? arrayType : symTable.semanticError;
```

Comment:
I wonder if we have to verify if this is valid in the spec, since it explicitly says

> The inherent type of the array value created is also byte[N].

Reply:
[#1150](https://github.com/ballerina-platform/ballerina-spec/issues/1150)